### ISSUE TITLE
buildah: builder does not work on btrfs file system

### ISSUE THREAD

**Issue by wiene** [hpc/charliecloud#793]
Trying to use `buildah` as builder on a system using btrfs as file system, I get the following error message:
```
$ ch-build -b buildah -t <tag> -f Dockerfile . 
building with: buildah
kernel does not support overlay fs: 'overlay' is not supported over btrfs at "/home/<user>/.local/share/containers/storage/overlay": backing file system is unsupported for this graph driver
```

**Comment by olifre** [comment]
Interestingly, the [BTRFS changelog](https://btrfs.wiki.kernel.org/index.php/Changelog) claims that OverlayFS should work with overlayfs on top since kernel 4.7, and overlayfs as lower filesystem since 4.15 (I'm not sure what exactly is meant with the phrasing). 

Checking out https://github.com/containers/libpod/issues/2172#issuecomment-455113462 , it seems that configuring buildah to use the btrfs storage driver should also work. Alternatively, it seems `fuse-overlayfs` can be used (see https://github.com/containers/buildah/issues/1304#issuecomment-458630736 ). 

Sadly, buildah does not seem to document this well :-(. Do any of these approaches yield success for you?

**Comment by wiene** [comment]
I'm running kernel 5.7.

Switching to the btrfs storage driver results in the following error:
```
building with: buildah
Unknown option .ignore_chown_errors
ERRO exit status 1
```
After the storage driver change, my file `~/.config/containers/storage.conf` looks like this:
```
[storage]
  driver = "btrfs"
  runroot = "/run/user/1000/containers"
  graphroot = "/home/<user>/.local/share/containers/storage"
  [storage.options]
    size = ""
    remap-uids = ""
    remap-gids = ""
    ignore_chown_errors = ""
    remap-user = ""
    remap-group = ""
    skip_mount_home = ""
    mount_program = "/usr/bin/fuse-overlayfs"
    mountopt = ""
    [storage.options.aufs]
      mountopt = ""
    [storage.options.btrfs]
      min_space = ""
      size = ""
    [storage.options.thinpool]
      autoextend_percent = ""
      autoextend_threshold = ""
      basesize = ""
      blocksize = ""
      directlvm_device = ""
      directlvm_device_force = ""
      fs = ""
      log_level = ""
      min_free_space = ""
      mkfsarg = ""
      mountopt = ""
      size = ""
      use_deferred_deletion = ""
      use_deferred_removal = ""
      xfs_nospace_max_retries = ""
    [storage.options.overlay]
      ignore_chown_errors = ""
      mountopt = ""
      mount_program = ""
      size = ""
      skip_mount_home = ""
    [storage.options.vfs]
      ignore_chown_errors = ""
    [storage.options.zfs]
      mountopt = ""
      fsname = ""
      size = ""
```
Strangely the error message remains unchanged even after removing all
```
    ignore_chown_errors = ""
```
lines.

**Comment by olifre** [comment]
Ah, that seems to be caused by hardcoding that option here:
https://github.com/hpc/charliecloud/blob/aa8d86e81dc543ed812eb99283e36a417682ac79/bin/ch-build#L86
Removing that line makes it work fine with the `btrfs` storage driver. 

So probably, we should find a way to not enforce passing that option via `ch-build` since it does not exist with all storage drivers. 

An alternative is the `vfs` driver, which also seems to work fine, and understands the option. 

I'm still confused as on why the `overlay` driver does not work. I can reproduce your problem. The error message is generated if this test:
https://github.com/containers/buildah/blob/9fd14e650469b623a634511ba93a1528574522e9/vendor/github.com/containers/storage/drivers/overlay/overlay.go#L360
fails.

**Comment by olifre** [comment]
> I'm still confused as on why the `overlay` driver does not work. I can reproduce your problem.

I tested:
```
buildah pull --debug docker.io/busybox
```
and got:
```
DEBU [graphdriver] trying provided driver "overlay" 
DEBU overlay test mount with multiple lowers failed mountfrom re-exec error: exit status 1: output: operation not permitted 
DEBU overlay test mount with a single lower failed mountfrom re-exec error: exit status 1: output: operation not permitted 
ERRO 'overlay' is not supported over btrfs at "/home/olifre/.local/share/containers/storage/overlay"
```
So I think this is not a `btrfs` issue at all, but just a very misleading error message: OverlayFS can not be used as unprivileged user. Searching around, other users report similar errors with extfs when running unprivileged. 
(note: If you want to reproduce, you may need to execute `/var/run/user/$(id -u)/overlay/overlay-false` first to clear the negative feature cache of `buildah`). 

Running as `root`, it works. 

So in short:
- We should find a solution for the `ignore_chown_errors` parameter, it does not work with all graphdrivers. 
- As actual workaround, the `vfs` graphdriver should work, and `btrfs` should work once we found a way not to specify that option unsupported by it.

**Comment by olifre** [comment]
Another issue: `buildah` itself also works fine for me even with `overlay` if I install `fuse-overlayfs` (which works unprivileged= and specify:
```
  [storage.options]
   mount_program = "/usr/bin/fuse-overlayfs"
```
but it breaks when used via `ch-build`, since `ch-build` overrides all `[storage.options]`. 

Does anybody have a good idea on how to fix this? 
We likely want to keep `.ignore_chown_errors` in general, but overriding any storage options the user may have set and seeing that the option is incompatible with some storage drivers (e.g. `btrfs`), that seems problematic.

**Comment by olifre** [comment]
Collecting everything, here's what I came up with. 

We currently have the following issues caused by the way `ch-build` uses `buildah`:

1. If the user has configured storage options, they are silently dropped (that's since setting options via CLI, as we do for `.ignore_chown_errors`, overrides *all* options from config). 
2. The graphdriver `btrfs`, which works unprivilegedly, is unusable (it does not understand that parameter). 
3. The graphdriver `overlay` with `fuse-overlayfs`, which works unprivilegedly, is unusable, since that needs to be enabled via a custom `mount-program` specification which is part of the `storage.options`, which we override in `ch-build`. 
4. Likely, also other graphdrivers are affected. 

Possible solutions:
1. We could drop `.ignore_chown_errors` completely, and document that this is a recommended storage option with `buildah` if the VFS graphdriver is used (which is the default driver, I think). 
2. We could try to parse `buildah info` leveraging `jq` or similar, and rebuild the options and conditionally add `.ignore_chown_errors` when we see the VFS graphdriver is used. 

After some more thinking, I tend to the first option, since I am unsure how stable `buildah info` output will be over time. 

@reidpr, what do you think?

**Comment by reidpr** [comment]
Thanks for your patience.

> If the user has configured storage options, they are silently dropped (that's since setting options via CLI, as we do for `.ignore_chown_errors`, overrides all options from config).

This is ludicrous, right? Isn't the standard thing that setting one option on the command line overrides that specific option from file config, not set all config to default except for what's on the command line?

> OverlayFS can not be used as unprivileged user.

Yes. You can't use overlayfs in an unprivileged user+mount namespace (except on Ubuntu).

> We could try to parse buildah info leveraging jq or similar, and rebuild the options and conditionally add .ignore_chown_errors when we see the VFS graphdriver is used.

I don't have a great deal of interest in implementing this. It seems complicated, and IMO it's a bug in Buildah that setting any storage option on the command line ignores the rest of the user's config.

> We could drop .ignore_chown_errors completely, and document that this is a recommended storage option with buildah if the VFS graphdriver is used (which is the default driver, I think).

This is more appealing, but again I hesitate to get too deep into explaining how to work around bugs in third-party software.

All that said, we do have a bug in `ch-build`: `$ignore_chown` ([link](https://github.com/hpc/charliecloud/blob/254fa07a2241ea0349179ea42471b1abdefafeea/bin/ch-build#L74)) is ignored. For `buildah-runc` and `buildah-setuid` builders, we should not add `--storage-opt`.

**Comment by olifre** [comment]
> This is ludicrous, right? Isn't the standard thing that setting one option on the command line overrides that specific option from file config, not set all config to default except for what's on the command line?

It is, I fully agree. 

Your explanation made me think about opening a bug against `buildah`, since indeed, that's rather ridiculous and unexpected behaviour. To do so, I checked the sources and the most up-to-date documentation... and the good news is, apparently, there is a way to get it to behave right:
https://github.com/containers/storage/blob/1745ea4e501f886899c96c5541f69f5c9046c4e6/store.go#L3534-L3536
So it seems setting the environment variable `STORAGE_OPTS` instead of specifying the command line parameter should append to the options instead of purging all options and replacing them with what is given on the command line. 

Still, that means graphdrivers not supporting that option will fail (`btrfs` for example). I don't have a good idea on that.

**Comment by reidpr** [comment]
So what if Charliecloud:

1. Actually respect `$ignore_chown`.
2. Put the option in `STORAGE_OPTS` instead of on the command line.
3. Document that not all graph drivers can use unprivileged build.
4. Document a pointer to Buildah graph driver docs (I'd need a link).

**Comment by reidpr** [comment]
So what I said about `$ignore_chown` is wrong: we do respect it, but its value is the argument to `--storage-opt .ignore_chown_errors`. Sounds like we need to instead just not provide `--storage-opt` unless the value is `true`. Previous comment edited.

**Comment by olifre** [comment]
@reidpr I fully agree, that sounds like a good plan :-). 
I'd add on "Document that not all graph drivers can use unprivileged build.": 
It's even a bit worse: Not all graph drivers can use unprivileged build everywhere (it might depend on file system, or downstream distribution kernel patches). 

Ideally, Buildah would document a list somewhere... 

And here we come to the main issue: I don't know of any good Buildah docs :-(. All I know about it is from trying it out with Charliecloud, and from experimenting with it. Sadly, their webpage does not provide any docs, just tutorials (or maybe I am not clever enough to find them), and even the manpage does not manage to list all available drivers. 

Is anybody subscribed to the issue clever enough to find docs, or are they just non-existent?

**Comment by reidpr** [comment]
`ch-build` is gone, so I don't think this is a Charliecloud issue any more. Closing.
